core/src/main/java/org/apache/iceberg/RESTTable.java

```diff
@@ -0,0 +1,71 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg;
+
+import java.util.Map;
+import java.util.function.Supplier;
+import org.apache.iceberg.catalog.TableIdentifier;
+import org.apache.iceberg.metrics.MetricsReporter;
+import org.apache.iceberg.rest.RESTClient;
+import org.apache.iceberg.rest.ResourcePaths;
+
+public class RESTTable extends BaseTable {
```

Comment:
I'm still a bit skeptical we need a whole `RESTTable` concept but I'll go through the code and offer a concrete approach

Reply:
So instead of `RESTTable`, I wonder if it would make sense to add 2 methods to `TableOperations`, 1 called `boolean performRemoteScan()` which defaults to `false` and another called `TableScan remoteScan()` which defaults to throwing an unsupported operation. `RESTTableOperations` would override this to return true based on the conditions for planning mode. 

I feel like this is logically consistent with the responsibility of `TableOperations` which is meant to abstract table metadata access.

Then the implementation of `BaseTable.newScan` would do a `if (ops.performRemoteScan()) {return ops.remoteScan()}`


Then `RESTTableOperations` itself can return the scan implementation you currently have, and it has all the context of the client, the headers, etc.

We expand the surface area of API of `RESTTableOperations` but again I think for this case it's logically consistent since we are talking about interacting with metadata, or lack thereof by delegating to a server to do that talking.

It also has the benefit of making the implementation a bit more flexible so that if some other catalog implementation has some concept like this they're not tied down to REST specific implementations.

Reply:
I think we'll also need to make sure whatever abstraction we introduce can work well with engine integrations. E.g. one open question is if and how this should fit in to Spark's Distributed planning model?

 I think to begin to keep it simple, we can just say that if remote planning is actually a requirement from the server, then we would make sure Spark does not do a distributed plan. But I think the model should be flexible though that in the future if we wanted to distribute the plan tasks as part of a distributed planning we could be able to do so.

Reply:
Thanks @amogh-jahagirdar for the suggestion. I think the reason why I opted for creating this `RESTTable` is that  in the future we may want to add more functionality to this class beyond just only the `newScan()`. 
For example I believe @geruh is working on revisiting the commits protocol, and my understanding is that we may overriding the implementation of the `newAppend()`, `newDelete()`, etc.

Reply:
Sounds good, I can understand the appeal of the `RESTTable` concept, especially with being able to override the operation implementations. I'm not very against it, just trying to avoid any unnecessary public classes being exposed, if there's somehow we can handle the redirection to the REST implementation internally of those things you mentioned. 

I'd say let's stick with what you have for now, as we write tests and see more integration code with engines like Spark, we can determine what's the right pattern.